File: src/main/java/com/dyonovan/neotech/common/tileentity/generator/TileFurnaceGenerator.java
package com.dyonovan.neotech.common.tileentity.generator;

import cofh.api.energy.EnergyStorage;
import cofh.api.energy.IEnergyProvider;
import cofh.api.energy.IEnergyReceiver;
import com.dyonovan.neotech.common.blocks.IExpellable;
import com.dyonovan.neotech.common.tileentity.BaseMachine;
import com.dyonovan.neotech.common.tileentity.InventoryTile;
import net.minecraft.block.Block;
import net.minecraft.block.material.Material;
import net.minecraft.init.Blocks;
import net.minecraft.init.Items;
import net.minecraft.item.*;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.network.NetworkManager;
import net.minecraft.network.Packet;
import net.minecraft.network.play.server.S35PacketUpdateTileEntity;
import net.minecraft.server.gui.IUpdatePlayerListBox;
import net.minecraft.tileentity.TileEntity;
import net.minecraft.util.EnumFacing;
import net.minecraft.util.EnumParticleTypes;
import net.minecraft.world.World;

import java.util.ArrayList;
import java.util.List;

import static net.minecraftforge.fluids.FluidContainerRegistry.drainFluidContainer;
import static net.minecraftforge.fluids.FluidContainerRegistry.isFilledContainer;

//TODO : Dont consume containers (buckets)
public class TileFurnaceGenerator extends BaseMachine implements IUpdatePlayerListBox, IEnergyProvider, IExpellable {

    public int currentBurnTime;
    public int totalBurnTime;
    public EnergyStorage energyRF;

    /**
     * Energy Creation per Tick
     */
    private static final int RF_TICK = 20;

    public static final int FUEL_SLOT = 0;

    public TileFurnaceGenerator() {
        energyRF = new EnergyStorage(10000);
        currentBurnTime = 0;
        totalBurnTime = 0;
        inventory = new InventoryTile(1);
    }

    @Override
    public void update() {
        if (!this.hasWorldObj()) return;
        World world = this.getWorld();
        if (world.isRemote) return;

        transferEnergy();

        if (currentBurnTime > 0 || (energyRF.getEnergyStored() < energyRF.getMaxEnergyStored() && inventory != null)) {
            if (currentBurnTime == 0) {
                totalBurnTime = getFuelValue(inventory.getStackInSlot(FUEL_SLOT));
                if (totalBurnTime == 0) return;
                currentBurnTime = 1;
                if (isFilledContainer(inventory.getStackInSlot(FUEL_SLOT)))
                        inventory.setStackInSlot(drainFluidContainer(inventory.getStackInSlot(FUEL_SLOT)),FUEL_SLOT);
                else if (inventory.getStackInSlot(FUEL_SLOT).stackSize == 1) inventory.setStackInSlot(null, FUEL_SLOT);
                else inventory.getStackInSlot(FUEL_SLOT).stackSize -= 1;
            }
            if (currentBurnTime > 0 && currentBurnTime < totalBurnTime) {
                energyRF.modifyEnergyStored(RF_TICK);
                currentBurnTime += 1;
            }
            if (currentBurnTime > 0 && currentBurnTime >= totalBurnTime) {
                currentBurnTime = 0;
                totalBurnTime = 0;
            }
            world.markBlockForUpdate(this.pos);
        }
    }

    private void transferEnergy() {
        List<EnumFacing> availDir = new ArrayList<>();
        if (energyRF.getEnergyStored() > 0) {
            for (EnumFacing dir : EnumFacing.VALUES) {
                TileEntity tile = getWorld().getTileEntity(this.pos.offset(dir));
                if (tile instanceof IEnergyReceiver) availDir.add(dir);
            }
        }
        if (availDir.size() <= 0) return;
        int availRF = Math.min(energyRF.getEnergyStored() / availDir.size() , energyRF.getMaxExtract() / availDir.size());
        for (EnumFacing dir : availDir) {
            TileEntity tile = getWorld().getTileEntity(this.pos.offset(dir));
            energyRF.extractEnergy(((IEnergyReceiver) tile).receiveEnergy(dir.getOpposite(), energyRF.extractEnergy(availRF, true), false), false);
        }
        getWorld().markBlockForUpdate(this.pos);
    }

    private int getFuelValue(ItemStack itemStack) {
        if (itemStack == null) return 0;

        Item item = itemStack.getItem();
        if (item instanceof ItemBlock && Block.getBlockFromItem(item) != Blocks.air) {
            Block block = Block.getBlockFromItem(item);
            if (block == Blocks.wooden_slab)
            {
                return 150;
            }

            if (block.getMaterial() == Material.wood)
            {
                return 300;
            }

            if (block == Blocks.coal_block)
            {
                return 16000;
            }
        }
        if (item instanceof ItemTool && ((ItemTool)item).getToolMaterialName().equals("WOOD")) return 200;
        if (item instanceof ItemSword && ((ItemSword)item).getToolMaterialName().equals("WOOD")) return 200;
        if (item instanceof ItemHoe && ((ItemHoe)item).getMaterialName().equals("WOOD")) return 200;
        if (item == Items.stick) return 100;
        if (item == Items.coal) return 1600;
        if (item == Items.lava_bucket) return 20000;
        if (item == Item.getItemFromBlock(Blocks.sapling)) return 100;
        if (item == Items.blaze_rod) return 2400;
        return net.minecraftforge.fml.common.registry.GameRegistry.getFuelValue(itemStack);
    }

    /*******************************************************************************************************************
     ************************************** Energy Functions ***********************************************************
     *******************************************************************************************************************/

    @Override
    public int extractEnergy(EnumFacing from, int maxExtract, boolean simulate) {
        return energyRF.extractEnergy(maxExtract, simulate);
    }

    @Override
    public int getEnergyStored(EnumFacing from) {
        return energyRF.getEnergyStored();
    }

    @Override
    public int getMaxEnergyStored(EnumFacing from) {
        return energyRF.getMaxEnergyStored();
    }

    @Override
    public boolean canConnectEnergy(EnumFacing from) {
        return true;
    }

    /*******************************************************************************************************************
     **************************************** Tile Functions ***********************************************************
     *******************************************************************************************************************/

    @Override
    public void readFromNBT(NBTTagCompound tag) {
        super.readFromNBT(tag);
        energyRF.readFromNBT(tag);
        inventory.readFromNBT(tag, this);
        currentBurnTime = tag.getInteger("CurrentBurnTime");
        totalBurnTime = tag.getInteger("TotalBurnTime");
    }

    @Override
    public void writeToNBT(NBTTagCompound tag) {
        super.writeToNBT(tag);
        energyRF.writeToNBT(tag);
        inventory.writeToNBT(tag);
        tag.setInteger("CurrentBurnTime", currentBurnTime);
        tag.setInteger("TotalBurnTime", totalBurnTime);
    }

    @Override
    public void spawnActiveParticles(double x, double y, double z) {
        List<EnumParticleTypes> particles = new ArrayList<>();
    }

    /*******************************************************************************************************************
     **************************************** Inventory Functions ******************************************************
     *******************************************************************************************************************/

    @Override
    public int[] getSlotsForFace(EnumFacing side) {
        return new int[] {0};
    }

    @Override
    public boolean canInsertItem(int index, ItemStack itemStackIn, EnumFacing direction) {
        return true;
    }

    @Override
    public boolean canExtractItem(int index, ItemStack stack, EnumFacing direction) {
        return false;
    }

    @Override
    public int getSizeInventory() {
        return 1;
    }

    @Override
    public boolean isItemValidForSlot(int index, ItemStack stack) {
        return getFuelValue(stack) > 0;
    }

    @Override
    public int getField(int id) {
        switch (id) {
            case 0:
                return currentBurnTime;
            case 1:
                return totalBurnTime;
            default:
                return 0;
        }
    }

    @Override
    public void setField(int id, int value) {
        switch (id) {
            case 0:
                currentBurnTime = value;
                break;
            case 1:
                totalBurnTime = value;
                break;
        }
    }

    @Override
    public int getFieldCount() {
        return 2;
    }

    @Override
    public void clear() {
        inventory.setStackInSlot(null, 0);
    }

    /*******************************************************************************************************************
     **************************************** Tile Functions ***********************************************************
     *******************************************************************************************************************/

    @Override
    public Packet getDescriptionPacket() {
        NBTTagCompound tag = new NBTTagCompound();
        this.writeToNBT(tag);
        return new S35PacketUpdateTileEntity(this.pos, 1, tag);
    }

    @Override
    public void onDataPacket(NetworkManager net, S35PacketUpdateTileEntity pkt) {
        readFromNBT(pkt.getNbtCompound());
    }
}

File: src/main/scala/com/dyonovan/neotech/common/tiles/machines/TileFurnaceGenerator.scala
package com.dyonovan.neotech.common.tiles.machines

import cofh.api.energy.{IEnergyReceiver, EnergyStorage}
import com.dyonovan.neotech.common.tiles.AbstractMachine
import net.minecraft.inventory.Container
import net.minecraft.item.ItemStack
import net.minecraft.tileentity.TileEntityFurnace
import net.minecraft.util.{EnumParticleTypes, EnumFacing}

/**
 * This file was created for NeoTech
 *
 * NeoTech is licensed under the
 * Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International License:
 * http://creativecommons.org/licenses/by-nc-sa/4.0/
 *
 * @author Dyonovan
 * @since August 13, 2015
 */
class TileFurnaceGenerator extends AbstractMachine {

    final val RF_TICK = 20

    override val energy = new EnergyStorage(100000)

    override def doWork(): Unit = {

        //Generate
        if (this.values.burnTime > 0) {
            this.values.burnTime -= 1
            energy.receiveEnergy(RF_TICK, false)
            worldObj.markBlockForUpdate(pos)
        } else if (energy.getEnergyStored < energy.getMaxEnergyStored && getStackInSlot(0) != null) {
            this.values.burnTime = (TileEntityFurnace.getItemBurnTime(getStackInSlot(0)) / 2.0).toInt
            this.values.currentItemBurnTime = this.values.burnTime
            decrStackSize(0, 1)
        }
        //Transfer
        if (energy.getEnergyStored > 0) {
            for (i <- EnumFacing.values()) {
                worldObj.getTileEntity(pos.offset(i)) match {
                    case tile: IEnergyReceiver =>
                        val want = tile.receiveEnergy(i.getOpposite, energy.getEnergyStored, true)
                        if (want > 0) {
                            val actual = energy.extractEnergy(want, false)
                            tile.receiveEnergy(i.getOpposite, actual, false)
                            worldObj.markBlockForUpdate(pos)
                        }
                    case _ =>
                }
            }
        }
    }
    /**
     * Get the output of the recipe
     * @param stack The input
     * @return The output
     */
    override def recipe(stack: ItemStack): ItemStack = null

    /**
     * Used to output the redstone single from this structure
     *
     * Use a range from 0 - 16.
     *
     * 0 Usually means that there is nothing in the tile, so take that for lowest level. Like the generator has no energy while
     * 16 is usually the flip side of that. Output 16 when it is totally full and not less
     *
     * @return int range 0 - 16
     */
    override def getRedstoneOutput: Int = Container.calcRedstoneFromInventory(this)

    override def initialSize: Int = 1

    override def canExtractItem(index: Int, stack: ItemStack, direction: EnumFacing): Boolean = false

    override def getSlotsForFace(side: EnumFacing): Array[Int] = Array[Int](0)

    override def receiveEnergy(from: EnumFacing, maxReceive: Int, simulate: Boolean): Int = 0

    override def extractEnergy(from: EnumFacing, maxExtract: Int, simulate: Boolean): Int = {
        val actual = energy.extractEnergy(maxExtract, simulate)
        worldObj.markBlockForUpdate(pos)
        actual
    }

    override def isItemValidForSlot(index: Int, stack: ItemStack): Boolean = {
        TileEntityFurnace.getItemBurnTime(stack) > 0
    }

    override def canInsertItem(index: Int, itemStackIn: ItemStack, direction: EnumFacing): Boolean = {
        index match {
            case 0 => TileEntityFurnace.getItemBurnTime (itemStackIn) > 0
            case _ => false
        }
    }

    override def spawnActiveParticles(x: Double, y: Double, z: Double): Unit = {
        worldObj.spawnParticle(EnumParticleTypes.FLAME, x, y, z, 0, 0, 0)
        worldObj.spawnParticle(EnumParticleTypes.SMOKE_NORMAL, x, y, z, 0, 0, 0)
    }
}
